### PR TITLE
perf(sql): introduce fast path for ordered and limited queries over single long column

### PR DESCRIPTION
**Pull request by puzpuzpuz**
Adds special factory for GROUP BY + ORDER BY long_column + LIMIT N queries. The factory executes a tight loop over the underlying `Map`(s), hence reduced top K, a.k.a. limited sorting, time.

The improvement in relevant clickbench queries may be seen below.

Screenshot from 2024-11-08 23-04-34

### CODE REVIEW

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 353 / 373 (94.64%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursor.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursor.java|36|45|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursorFactory.java|31|36|86.11%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursor.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMaxHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMinHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered2MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMapCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/ShardedMapCursor.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeRecord.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered8MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered4MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMap.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapFixedSizeCursor.java|38|38|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursorFactory.java|1|1|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 354 / 373 (94.91%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursor.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursor.java|36|45|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursorFactory.java|31|36|86.11%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMaxHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMinHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered2MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMapCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/ShardedMapCursor.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursor.java|8|8|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeRecord.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered8MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered4MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMap.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapFixedSizeCursor.java|38|38|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursorFactory.java|1|1|100.00%|

**Comment by puzpuzpuz** [comment]
@nwoolmer thanks for the review!

**Review by bluestreak01 — CHANGES_REQUESTED** [review]

File: core/src/main/java/io/questdb/cairo/map/OrderedMapFixedSizeCursor.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+import io.questdb.cairo.sql.Function;
+import io.questdb.cairo.sql.Record;
+import io.questdb.cairo.sql.SqlExecutionCircuitBreaker;
+import io.questdb.std.DirectLongLongHeap;
+import io.questdb.std.bytes.Bytes;
+
+class OrderedMapFixedSizeCursor implements OrderedMapCursor {
+    private final long entrySize;
+    private final OrderedMap map;
```
[the remark this answers is not in the document]
Comment (puzpuzpuz) [reply core/src/main/java/io/questdb/cairo/map/OrderedMapFixedSizeCursor.java]: Indeed they must not be ignored, but when refactoring they're very easy to overlook, so the only reliable way to get rid of them is to introduce a CI check for the PRs.

Fixed in c3bede5156d08e6aaec3b5246d446784b084f07a

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 358 / 383 (93.47%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/map/Unordered8MapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered4MapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered2MapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursor.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeRecord.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursor.java|36|45|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursorFactory.java|31|36|86.11%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursor.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMaxHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMinHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapFixedSizeRecord.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered2MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMapCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/ShardedMapCursor.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered8MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered4MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMap.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapFixedSizeCursor.java|36|36|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursorFactory.java|1|1|100.00%|

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 358 / 383 (93.47%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/map/Unordered8MapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered4MapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered2MapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursor.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeRecord.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursor.java|36|45|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursorFactory.java|31|36|86.11%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursor.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMaxHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMinHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapFixedSizeRecord.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered2MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMapCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/ShardedMapCursor.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered8MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered4MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMap.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapFixedSizeCursor.java|36|36|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursorFactory.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Comment by glasstiger** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 358 / 383 (93.47%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/cairo/map/Unordered8MapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered4MapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered2MapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMapRecord.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursor.java|0|1|00.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeRecord.java|3|4|75.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursor.java|36|45|80.00%|
|:large_blue_circle:|io/questdb/griffin/engine/orderby/LongTopKRecordCursorFactory.java|31|36|86.11%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursor.java|7|8|87.50%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMaxHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/std/DirectLongLongMinHeap.java|52|54|96.30%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapFixedSizeRecord.java|1|1|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapVarSizeCursor.java|39|39|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered2MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMapCursor.java|12|12|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/groupby/GroupByRecordCursorFactory.java|5|5|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/ShardedMapCursor.java|3|3|100.00%|
|:large_blue_circle:|io/questdb/griffin/engine/table/AsyncGroupByRecordCursorFactory.java|4|4|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMap.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered8MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/griffin/SqlCodeGenerator.java|9|9|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/Unordered4MapCursor.java|16|16|100.00%|
|:large_blue_circle:|io/questdb/std/Numbers.java|6|6|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/UnorderedVarcharMap.java|7|7|100.00%|
|:large_blue_circle:|io/questdb/cairo/map/OrderedMapFixedSizeCursor.java|36|36|100.00%|
|:large_blue_circle:|io/questdb/cairo/sql/RecordCursorFactory.java|1|1|100.00%|

**Review by bluestreak01 — APPROVED** [review]

**Pull request merged**